### PR TITLE
feat: Improve parser with better notation and error messages

### PR DESCRIPTION
**Pull request by smalewski**
My idea is to support the following notations:

Products:
```
A -> B
forall (x : A), B -> C x
∀ (x y z : A) (n m : B), C x m
```

Lambdas:
```
fun x : A . B
λ (x : A) . B
λ (x y : A) (z : B) . C
```

Universes:
```
Type0
Type 1
□₂
```

Unknown:
```
?0
?₂
```

### CODE REVIEW

File: lib/parsing/lex_and_parse.ml [hunk 1]
```diff
@@ -1,18 +1,78 @@
...
+  | Lexer.SyntaxError msg ->
+      let error_msg = asprintf "%s" msg in
       Error error_msg
-  | Parser.Error ->
-      let error_msg = asprintf "%s: syntax error@." (print_error_position lexbuf) in
-      Error error_msg
\ No newline at end of file
+
+let term_of_string text =
+  parse_term text
```
Comment (TDiazT) [inline lib/parsing/lex_and_parse.ml:78]: Do we need it?

File: lib/parsing/lex_and_parse.mli [hunk 2]
```diff
@@ -3,4 +3,6 @@
 
 (** Given a lex buffer to read a GCIC term from, parse the
 term and return the AST if successful *)
-val parse_term : Lexing.lexbuf -> (Ast.term, string) result
\ No newline at end of file
+val parse_term : string -> (Ast.term, string) result
+
+val term_of_string : string -> (Ast.term, string) result
```
[the remark this answers is not in the document]
Comment (smalewski) [reply lib/parsing/lex_and_parse.mli]: In fact the function `term_of_string` is not needed.
Solved in 6c971c42ceb313283f5968fa1f967cd4c33400d9

File: lib/parsing/lexer.ml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  | ')'            -> RPAREN 
+  | '='            -> EQUAL 
+  | ':'            -> COLON 
+  | '.'            -> DOT 
+  | ','            -> COMMA 
+  | whitespace     -> token lexbuf
+  | newline        -> Sedlexing.new_line lexbuf; token lexbuf
+  | eof            -> EOF 
+  | any            -> raise (SyntaxError ("Unexpected char: " ^ lexeme lexbuf))
+  | _              -> raise (SyntaxError "Unexpected char")
```
[the remark this answers is not in the document]
Comment (smalewski) [reply lib/parsing/lexer.ml]: The wild case is redundant, but needed for exhaustiveness.
Solved in 6c971c42ceb313283f5968fa1f967cd4c33400d9

File: lib/parsing/parser.mly [hunk 4]
```diff
@@ -11,19 +11,16 @@
...
-%token  KWD_UNKNOWN
+%token <int> INT
+%token <string> ID
+%token COLON DOT COMMA ARROW EQUAL
+%token LPAREN RPAREN
+%token KWD_UNIVERSE KWD_LAMBDA KWD_UNKNOWN KWD_FORALL
+%token KWD_LET KWD_IN
 %token EOF
 
+%on_error_reduce term
```
[the remark this answers is not in the document]
Comment (smalewski) [reply lib/parsing/parser.mly:22]: I added a comment in 6c971c42ceb313283f5968fa1f967cd4c33400d9

File: lib/parsing/dune [hunk 5]
```diff
@@ -1,9 +1,52 @@
 (menhir
- (modules parser))
-
-(ocamllex lexer)
+ (modules parser)
+ (flags --table --dump --strict --explain))
```
[the remark this answers is not in the document]
Comment (smalewski) [reply lib/parsing/dune:7]: Done.

File: lib/parsing/dune [hunk 6]
```diff
@@ -1,9 +1,52 @@
...
 (library
  (name parsing)
- (modules parser lexer ast lex_and_parse)
- (libraries common))
+ (modules parser lexer ast lex_and_parse parserMessages)
+ (libraries common sedlex menhirLib)
+ (preprocess
+  (pps sedlex.ppx)))
+
+;; This section deals with the .messages file.
```
[the remark this answers is not in the document]
Comment (smalewski) [reply lib/parsing/dune:18]: Done.

File: lib/parsing/lex_and_parse.ml (hunk 1 above)
Comment (smalewski) [reply lib/parsing/lex_and_parse.ml]: Done

**Pull request merged**